src/overlays/actors/ovl_En_Bbfall/z_en_bbfall.c

```diff
@@ -37,140 +47,717 @@ const ActorInit En_Bbfall_InitVars = {
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Bbfall/func_808C080C.s")
+/**
+ * This maps a given limb based on its limbIndex to its appropriate index
+ * in the bodyPartsPos/Velocity arrays. An index of -1 indicates that the
+ * limb is not part of the bodyParts arrays.
+ */
+static s8 sLimbIndexToBodyPartsIndex[] = {
+    -1, -1, -1, -1, 0, -1, -1, -1, 1, -1, -1, -1, -1, 2, -1, 3,
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Bbfall/EnBbfall_Draw.s")
+/**
+ * The last element of the bodyParts arrays is a duplicate of the cranium
+ * limb, which is then offset by a certain amount. There is no display list
+ * associated with this, so it is only used for effects.
+ */
+static Vec3f sDuplicateCraniumBodyPartOffset = { 1000.0f, -700.0f, 0.0f };
+
+void EnBbfall_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnBbfall* this = THIS;
+    s32 i;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    SkelAnime_Init(globalCtx, &this->skelAnime, &gBubbleSkel, &gBubbleFlyingAnim, this->jointTable, this->morphTable,
+                   BUBBLE_LIMB_MAX);
+    CollisionCheck_SetInfo(&this->actor.colChkInfo, &sDamageTable, &sColChkInfoInit);
+    Collider_InitAndSetJntSph(globalCtx, &this->collider, &this->actor, &sJntSphInit, this->colliderElements);
+    ActorShape_Init(&this->actor.shape, 1500.0f, ActorShadow_DrawCircle, 35.0f);
+    this->timer = 0;
+    EnBbfall_SetupWaitForPlayer(this);
+    Actor_SetFocus(&this->actor, 0.0f);
+    for (i = 0; i < ARRAY_COUNT(this->colliderElements); i++) {
+        this->collider.elements[i].dim.worldSphere.radius = this->collider.elements[i].dim.modelSphere.radius;
+    }
+}
+
+void EnBbfall_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    EnBbfall* this = THIS;
+
+    Collider_DestroyJntSph(globalCtx, &this->collider);
+}
+
+void EnBbfall_Freeze(EnBbfall* this) {
+    this->drawDmgEffType = ACTOR_DRAW_DMGEFF_FROZEN_NO_SFX;
+    this->drawDmgEffScale = 0.4f;
+    this->drawDmgEffFrozenSteamScale = 0.6f;
+    this->timer = 80;
+    this->drawDmgEffAlpha = 1.0f;
+    this->actor.flags &= ~ACTOR_FLAG_200;
+    Actor_SetColorFilter(&this->actor, 0x4000, 255, 0, 80);
+}
+
+void EnBbfall_Thaw(EnBbfall* this, GlobalContext* globalCtx) {
+    if (this->drawDmgEffType == ACTOR_DRAW_DMGEFF_FROZEN_NO_SFX) {
+        this->drawDmgEffType = ACTOR_DRAW_DMGEFF_FIRE;
+        this->drawDmgEffAlpha = 0.0f;
+        Actor_SpawnIceEffects(globalCtx, &this->actor, this->bodyPartsPos, ARRAY_COUNT(this->bodyPartsPos), 2, 0.2f,
+                              0.15f);
+        this->actor.flags |= ACTOR_FLAG_200;
+    }
+}
+
+s32 EnBbfall_IsTouchingLava(EnBbfall* this, GlobalContext* globalCtx) {
+    if (!SurfaceType_IsWallDamage(&globalCtx->colCtx, this->actor.floorPoly, this->actor.floorBgId)) {
+        u32 floorType = func_800C99D4(&globalCtx->colCtx, this->actor.floorPoly, this->actor.floorBgId);
+
+        if ((floorType == 2) || (floorType == 3) || (floorType == 9)) {
```

Comment:
Just random thoughts here is this checking for different lava floor types? Could this function be `EnBbfall_IsTouchingFloor`?

Reply:
It's definitely checking for lava floor types. The goal of this function is to return true if the floor is the type that the Red Bubble can "sink into". If I make the function return true no matter what, then this happens:

https://user-images.githubusercontent.com/12245827/162553551-46f07a00-46fc-48d0-8bea-d8506a68dc3e.mp4

I'll add a comment above the function explaining this.